web/html/src/components/table/AsyncDataHandler.js

```diff
@@ -0,0 +1,318 @@
+  initialSortColumnKey?: string,
+  /** 1 for ascending, -1 for descending */
+  initialSortDirection?: number,
+  /** the React Object that contains the filter search field */
+  searchField?: React.Element<typeof SearchField>,
+  /** the initial number of how many row-per-page to show. If it's 0 table header and footer are hidden */
+  initialItemsPerPage?: number,
+  /** enables item selection */
+  selectable: boolean,
+  /** the handler to call when the table selection is updated. If not provided, the select boxes won't be rendered */
+  onSelect?: (items: Array<any>) => void,
+  /** the identifiers for selected items */
+  selectedItems?: Array<any>,
+  /** The message which is shown when there are no rows to display */
+  emptyText?: string,
+  /** The message which is shown when the data is loading */
+  loadingText?: string,
+  /** Children node in the table */
+  children: (args: ChildrenArgsProps) => React.Node,
+};
+
+type State = {
+  data: Array<any>,
+  currentPage: number,
+  itemsPerPage: number,
+  totalItems: number,
+  criteria?: string,
+  sortColumnKey: string | null,
+  sortDirection: number,
+  selectedItems: Array<any>,
+  selectable: boolean,
+  loading: boolean,
+  endpoint: ApiDataEndpoint
+};
+
+export class AsyncDataHandler extends React.Component<Props, State> {
+  static defaultProps = {
+    selectable: false,
+    columns: [],
+  };
+
+  static contextType = StatePersistedContext;
+
+  constructor(props: Props) {
+    super(props);
+    this.state = {
+      data: [],
+      currentPage: 1,
+      itemsPerPage: this.props.initialItemsPerPage || 15,
+      totalItems: 0,
+      criteria: undefined,
+      sortColumnKey: this.props.initialSortColumnKey || null,
+      sortDirection: this.props.initialSortDirection || 1,
+      selectedItems: this.props.selectedItems || [],
+      selectable: this.props.selectable,
+      loading: false,
+      endpoint: new ApiDataEndpoint(new URL(this.props.resource, window.location.origin))
+    };
+  };
+
+  getData() {
+    const currPage = (this.state.currentPage - 1) * this.state.itemsPerPage + 1;
+    const pageControl = new PageControl(currPage, this.state.itemsPerPage,
+      this.state.criteria, this.state.sortColumnKey, this.state.sortDirection);
+
+    this.setState({loading: true}, () => {
+      this.state.endpoint.get(promise => {
+        promise.then(data => this.updateData(data))
+          .finally(() => this.setState({loading: false}));
+        }, pageControl);
+    });
+  }
+
+  updateData({items, total}: {items: Array<any>, total: number}) {
+    this.setState({data: items, totalItems: total}, () => {
+      const lastPage = this.getLastPage();
+      if (this.state.currentPage > lastPage) {
+        this.setState({currentPage: lastPage});
+      }
+    });
+  }
+
+  UNSAFE_componentWillReceiveProps(nextProps: Props) {
+    this.setState({
+        selectedItems: nextProps.selectedItems || [],
+        selectable: Boolean(nextProps.selectable)
+    });
+  };
```

Comment:
I know it's a copy/paste from `TableDataHandler` but would be nice to get rid of the `UNSAFE_*` functions in that code